src/rules/no-incompatible-branch-types.coffee

```diff
@@ -0,0 +1,63 @@
+ts = require 'typescript'
+
+# getSourceFileOfNode = (tsNode) ->
+#   while node and node.kind isnt ts.SyntaxKind.SourceFile
+#     node = node.parent
+#   node
+
+module.exports =
+  meta:
+    docs:
+      description:
+        'Flag incompatible early-return types in branch()/branchPure()'
+      category: 'Possible Errors'
+      recommended: yes
+    schema: []
+
+  create: (context) ->
+    {parserServices} = context
+    return {} unless parserServices?.hasFullTypeInformation
```

Comment:
I mimicked this pattern from `eslint-plugin-reanimated` which seems like it can replace the existing `isTypescript()` sniffing utility from this repo